File: tests/observability/config.test.py
"""Twin of observability-config.test.ts.

The loader reads already-populated ``os.environ`` (it does NOT re-parse .env --
that's the bridges' job, PR #416), so this tests precedence, truthy parsing,
and malformed-override fallback.
"""

from __future__ import annotations

import json
import os
import sys
import tempfile
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "src"))

from observability import config as cfgmod  # noqa: E402

KNOBS = (
    "SUTANDO_TENANT_ID",
    "SUTANDO_TENANT_MODE",
    "SUTANDO_METERING_ENABLED",
    "SUTANDO_METERING_ENDPOINT",
)


class ObservabilityConfigTest(unittest.TestCase):
    def setUp(self) -> None:
        self._saved = {k: os.environ.pop(k, None) for k in KNOBS}
        self._tmp = tempfile.TemporaryDirectory()
        self.ws = Path(self._tmp.name)

    def tearDown(self) -> None:
        for k, v in self._saved.items():
            if v is None:
                os.environ.pop(k, None)
            else:
                os.environ[k] = v
        self._tmp.cleanup()

    def test_defaults_only(self) -> None:
        self.assertEqual(cfgmod.load_observability_config(self.ws), cfgmod.OBSERVABILITY_DEFAULTS)

    def test_env_beats_defaults(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "acct_123"
        os.environ["SUTANDO_TENANT_MODE"] = "managed"
        os.environ["SUTANDO_METERING_ENABLED"] = "true"
        os.environ["SUTANDO_METERING_ENDPOINT"] = "https://meter.example"
        cfg = cfgmod.load_observability_config(self.ws)
        self.assertEqual(cfg["tenant"]["id"], "acct_123")
        self.assertEqual(cfg["tenant"]["mode"], "managed")
        self.assertTrue(cfg["metering"]["enabled"])
        self.assertEqual(cfg["metering"]["endpoint"], "https://meter.example")

    def test_metering_enabled_truthy_parse(self) -> None:
        os.environ["SUTANDO_METERING_ENABLED"] = "no"
        self.assertFalse(cfgmod.load_observability_config(self.ws)["metering"]["enabled"])
        os.environ["SUTANDO_METERING_ENABLED"] = "on"
        self.assertTrue(cfgmod.load_observability_config(self.ws)["metering"]["enabled"])

    def test_workspace_override_wins(self) -> None:
        os.environ["SUTANDO_TENANT_MODE"] = "managed"
        (self.ws / "config").mkdir(parents=True, exist_ok=True)
        (self.ws / "config" / "observability.json").write_text(
            json.dumps({"tenant": {"mode": "byok"}, "observability": {"sampling": {"trace": 0.5}}})
        )
        cfg = cfgmod.load_observability_config(self.ws)
        self.assertEqual(cfg["tenant"]["mode"], "byok")  # workspace overrides env
        self.assertEqual(cfg["observability"]["sampling"]["trace"], 0.5)
        self.assertEqual(cfg["metering"]["batchMax"], 100)  # untouched key kept

    def test_malformed_override_falls_back(self) -> None:
        (self.ws / "config").mkdir(parents=True, exist_ok=True)
        (self.ws / "config" / "observability.json").write_text("{ not valid json")
        cfg = cfgmod.load_observability_config(self.ws)
        self.assertEqual(cfg, cfgmod.OBSERVABILITY_DEFAULTS)


if __name__ == "__main__":
    unittest.main()

File: tests/observability/config.test.ts
import { describe, it, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, mkdirSync, writeFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
	loadObservabilityConfig,
	OBSERVABILITY_DEFAULTS,
} from '../../src/observability/config.js';

const KNOBS = ['SUTANDO_TENANT_ID', 'SUTANDO_TENANT_MODE', 'SUTANDO_METERING_ENABLED', 'SUTANDO_METERING_ENDPOINT'];
let saved: Record<string, string | undefined>;
let ws: string;

beforeEach(() => {
	saved = {};
	for (const k of KNOBS) {
		saved[k] = process.env[k];
		delete process.env[k];
	}
	ws = mkdtempSync(join(tmpdir(), 'obscfg-'));
});

afterEach(() => {
	for (const k of KNOBS) {
		if (saved[k] === undefined) delete process.env[k];
		else process.env[k] = saved[k];
	}
	rmSync(ws, { recursive: true, force: true });
});

describe('kernel/config/observability-config', () => {
	it('clean env + no override → equals OBSERVABILITY_DEFAULTS', () => {
		assert.deepEqual(loadObservabilityConfig({ workspace: ws }), OBSERVABILITY_DEFAULTS);
	});

	it('env knobs beat the in-code defaults', () => {
		process.env.SUTANDO_TENANT_ID = 'acct_123';
		process.env.SUTANDO_TENANT_MODE = 'managed';
		process.env.SUTANDO_METERING_ENABLED = 'true';
		process.env.SUTANDO_METERING_ENDPOINT = 'https://meter.example';
		const cfg = loadObservabilityConfig({ workspace: ws });
		assert.equal(cfg.tenant.id, 'acct_123');
		assert.equal(cfg.tenant.mode, 'managed');
		assert.equal(cfg.metering.enabled, true);
		assert.equal(cfg.metering.endpoint, 'https://meter.example');
	});

	it('SUTANDO_METERING_ENABLED parses only truthy tokens', () => {
		process.env.SUTANDO_METERING_ENABLED = 'no';
		assert.equal(loadObservabilityConfig({ workspace: ws }).metering.enabled, false);
		process.env.SUTANDO_METERING_ENABLED = 'on';
		assert.equal(loadObservabilityConfig({ workspace: ws }).metering.enabled, true);
	});

	it('workspace override wins over env (documented order)', () => {
		process.env.SUTANDO_TENANT_MODE = 'managed';
		mkdirSync(join(ws, 'config'), { recursive: true });
		writeFileSync(
			join(ws, 'config', 'observability.json'),
			JSON.stringify({ tenant: { mode: 'byok' }, observability: { sampling: { trace: 0.5 } } }),
		);
		const cfg = loadObservabilityConfig({ workspace: ws });
		assert.equal(cfg.tenant.mode, 'byok'); // workspace overrides the env knob
		assert.equal(cfg.observability.sampling.trace, 0.5);
		// untouched keys keep their resolved value (metering still defaulted)
		assert.equal(cfg.metering.batchMax, 100);
	});

	it('malformed workspace override → warns and falls back, never throws', () => {
		mkdirSync(join(ws, 'config'), { recursive: true });
		writeFileSync(join(ws, 'config', 'observability.json'), '{ not valid json');
		const cfg = loadObservabilityConfig({ workspace: ws });
		assert.deepEqual(cfg, OBSERVABILITY_DEFAULTS);
	});
});
